lib/spack/spack/environment/environment.py

```diff
@@ -823,18 +859,10 @@ def _re_read(self):
         """Reinitialize the environment object."""
         self.clear(re_read=True)
         self.manifest = EnvironmentManifestFile(self.path)
-        self._read(re_read=True)
+        self._read()
```

Comment:
```suggestion
        with manifest_config(self.manifest):
            self._read()
```
or if we move the manifest-config-related functions to be manifest methods, this could be:
```
        with self.manifest.use_config():
            self._read()
```

Reply:
We should also think about whether this needs a transaction, I think it does. We call it with a ReadTransaction when we call `_read` in the constructor. That looks like a preexisting bug to me.

Reply:
I did not see any difference using the new context manager in `_re_read` in the experiments I conducted.  But I'll revisit as maybe I missed something.

Reply:
Went ahead and made the change via a refactor of the handling so consistent with the lock, construct, and read of the manifest.